src/posts/2020-12-21-ebpf-updates/index.md

```diff
@@ -0,0 +1,393 @@
+  [link](https://lore.kernel.org/bpf/20201202172516.3483656-1-sdf@google.com/t/#u))
+
+* Add a new helper `bpf_ktime_get_coarse_ns()` to get a timestamp using the
+  `CLOCK_MONOTONIC_COARSE`, which is less accurate but more performant than
+  `CLOCK_MONOTONIC` used by `bpf_ktime_get_ns()`.
+  (Dmitrii Banshchikov,
+  [link](https://lore.kernel.org/bpf/20201117184549.257280-1-me@ubique.spb.ru/t/#u))
+
+This was followed by
+[a third pull request](https://lore.kernel.org/bpf/20201214214316.20642-1-daniel@iogearbox.net/),
+for the same tree and cycle, and for which you will find the highlights below:
+
+* Expose `bpf_sk_storage_get()` and `bpf_sk_storage_delete()` helpers to
+  iterator programs, for example to initialize or delete selected value from a
+  socket local storage from an eBPF iterator.
+  (Florent Revest,
+  [link](https://lore.kernel.org/bpf/20201204113609.1850150-3-revest@google.com/t/#u))
+
+* Add AF\_XDP selftests based on veth devices, for both SKB and native modes,
+  as part of the eBPF selftests suite.
+  (Weqaar Janjua,
+  [link](https://lore.kernel.org/bpf/20201207215333.11586-1-weqaar.a.janjua@intel.com/t/#u))
+
+* Update libbpf function `bpf_program__set_attach_target()` so it supports
+  finding BTF-based kernel attach targets (such as fentry, fexit, BTF-based
+  tracepoints, etc.). Like for other hooks, the attach points can be passed
+  through the ELF section name where the eBPF program is placed, with the
+  `SEC()` macro.
+  (Andrii Nakryiko,
+  [link](https://lore.kernel.org/bpf/20201211215825.3646154-1-andrii@kernel.org/t/#u))
+
+* Permit pointers on stack for helper calls in the verifier, if the user has
+  sufficient permissions. This addresses an issue where the verifier would
+  wrongly reject some programs.
+  (Yonghong Song,
+  [link](https://lore.kernel.org/bpf/20201210013348.943623-1-yhs@fb.com/t/#u))
+
+* Add new libbpf API function to retrieve an eBPF ring buffer epoll file
+  descriptor. This is to help with the migration from perf buffer to eBPF ring
+  buffer.
+  (Brendan Jackman,
+  [link](https://lore.kernel.org/bpf/20201214113812.305274-1-jackmanb@google.com/t/#u))
+
+Kernel 5.10 was released on the 13th of December, bringing a number of eBPF
+novelties. eBPF iterators with preload at boot time, or sleep-able eBPF
+programs, among others. Here I am deferring to [the list on the
+Kernelnewbies](https://kernelnewbies.org/Linux_5.10#Tracing.2C_perf_and_BPF)
+website for the detailed set of additions.
+
+## Did You Know?: CO-RE
+
+CO-RE (_Compile Once, Run Everywhere_) is a mechanism used with eBPF to ensure
+portability of the programs, mainly those intended for tracing. It addresses
+the issue that arises when a given structure is modified between two kernel
+versions. Tracing programs may attempt to access to a field from a given
+structure by reading at a specific offset in that structure. But if the length
+of the structure, or the order of its member fields, for example, is modified
```

Comment:
I'd drop the `for example` because (1) I can't think of other cases and (2) it becomes a bit heavy in parenthetic clauses (& commas).

Reply:
For 1), from Andrii's post:
> Different kernel versions will have struct fields shuffled around inside a struct, or even moved into a new inner struct. Fields can be renamed or removed, their types changed, either into some trivially-compatible ones or completely different ones. Structs and other types can get renamed, or they can be conditionally compiled out (depending on kernel configuration), or just plain removed between kernel versions.

Reply:
Ah, right. I forgot about types :+1: